pkg/util/slices.go

```diff
@@ -52,23 +49,42 @@ func CombineUnique[T constraints.Ordered](a, b []T) []T {
 	return a
 }
 
-// Filter returns a new slice that only contains elements from collection that
-// satisfy predicate.
+// Filter appends elements from collection that satisfy predicate to out and
+// returns it. Accepting out as a parameter allows callers to fine tune
+// allocation performance.
 //
-//	// Filter in place
-//	numbers = Filter(numbers, isEven)
-//	// Filter into a new slice
-//	odds := Filter(numbers, isEven)
-func Filter[T any](collection []T, predicate func(T) bool) []T {
-	i := 0
-	out := make([]T, len(collection))
-	for j := range collection {
-		if predicate(collection[j]) {
-			out[i] = collection[j]
-			i++
+//	// As a special case, if out == nil, a new slice the size of collection will be
+//	// allocated and returned. The following two statements are equivalent.
+//	even := Filter(numbers, isEven, nil)
+//	even := Filter(numbers, isEven, make([]int, 0, len(numbers))
+//
+//	// Force append to resize if any matching elements are found.
+//	_ := Filter(numbers, isEven, []int{})
+//
+//	// Zero allocations, reuses the original slice.
+//	numbers = Filter(numbers, isEven, numbers[:0])
+//
+//	// Exactly one allocation if we know/guessed the number of elements that
+//	// satisfy the predicate (e.g. we did a "first pass" over it).
+//	even := Filter(numbers, isEven, make([]int, 0, 10))
+//
+//	// We optimistically halve the slice size, but allocate 1 extra time if
+//	// it doesn't work out.
+//	even := Filter(numbers, isEven, make([]int, 0, len(numbers)/2)
+func Filter[T any](collection []T, predicate func(T) bool, out []T) []T {
```

Comment:
Consider making this work with any `~[]T` type, i.e. any types that have a slice as their underlying type.

```
func Filter[S ~[]T, T any](collection S, predicate func(T) bool, out S) S {
	...
}
```

See https://go.dev/blog/deconstructing-type-parameters for example of how this all works.